parser/openfoodfacts_taxonomy_parser/parser/parser.py

```diff
@@ -1,158 +1,213 @@
+            f"Created {count} 'is_before' links in {timeit.default_timer() - start_time} seconds"
+        )
+
+        if count != len(previous_links):
+            self.parser_logger.error(
+                f"Created {count} 'is_before' links, {len(previous_links)} links expected"
+            )
+
+    def _create_child_links(
+        self, child_links: list[ChildLink], entry_nodes: list[NodeData], project_label: str
+    ):
+        """Create the 'is_child_of' relations between nodes"""
         self.parser_logger.info("Creating 'is_child_of' links")
+        start_time = timeit.default_timer()
+
+        node_ids = set([node.id for node in entry_nodes])
+        normalised_child_links = []
+        unnormalised_child_links = []
         for child_link in child_links:
-            child_id = child_link["id"]
-            parent = child_link["parent_id"]
-            lc, parent_id = parent.split(":")
-            query = f""" MATCH (p:{multi_label}:ENTRY) WHERE $parent_id IN p.tags_ids_""" + lc
-            query += f"""
-                MATCH (c:{multi_label}) WHERE c.id= $child_id
-                CREATE (c)-[r:is_child_of]->(p)
-                RETURN r
-            """
-            result = self.session.run(query, parent_id=parent_id, child_id=child_id)
-            if not result.value():
-                self.parser_logger.warning(
-                    f"parent not found for child {child_id} with parent {parent_id}"
-                )
-
-    def _create_fulltext_index(self, taxonomy_name: str, branch_name: str):
-        """Create indexes for search"""
-        project_name = self._get_project_name(taxonomy_name, branch_name)
-        query = (
-            f"""CREATE FULLTEXT INDEX {project_name+'_SearchIds'} IF NOT EXISTS
-            FOR (n:{project_name}) ON EACH [n.id]\n"""
-            + """
-            OPTIONS {indexConfig: {`fulltext.analyzer`: 'keyword'}}"""
+            if child_link["parent_id"] in node_ids:
+                normalised_child_links.append(child_link)
+            else:
+                unnormalised_child_links.append(child_link)
+
+        normalised_query = f"""
+            UNWIND $normalised_child_links as child_link
+            MATCH (p:{project_label}) USING INDEX p:{project_label}(id)
+            WHERE p.id = child_link.parent_id
+            MATCH (c:{project_label}) USING INDEX c:{project_label}(id)
+            WHERE c.id = child_link.id
+            CREATE (c)-[relations:is_child_of]->(p)
+            WITH relations
+            UNWIND relations AS relation
+            RETURN COUNT(relation)
+        """
+
+        language_codes = set()
+        for child_link in unnormalised_child_links:
+            lc, parent_id = child_link["parent_id"].split(":")
+            language_codes.add(lc)
+            child_link["parent_id"] = parent_id
+
+        parent_id_query = " OR ".join(
+            [f"child_link.parent_id IN p.tags_ids_{lc}" for lc in language_codes]
         )
```

Comment:
This way you de-corelate the information between language and id, it seems dangerous to me, it's not sure that you won't match the wrong parent ! More over it's not the best for performances.

I would either run a query per language or change child_link to have language and id separated so that you can write: `child_link.parent_id IN p.tags_ids_{child_link.lc}` (this can be done by changing to dict())

Reply:
Yes thanks for finding the bug!